Stop changes to a dupped `ActionController::Parameters` mutating the original

timrogers opened the pull request:
When `ActionController::Parameters` is duplicated with `#dup`, it doesn't create a duplicate of the instance variables (e.g. `@parameters`) but rather maintains the reference (see http://ruby-doc.org/core-2.3.1/Object.html). Given that the parameters object is often manipulated as if it were a hash (e.g. with `#delete` and similar methods), this leads to unexpected behaviour, like the following:

```
params = ActionController::Parameters.new(foo: "bar")
duplicated_params = params.dup
duplicated_params.delete(:foo)

params == duplicated_params
```

This fixes the bug by defining a private `#initialize_copy` method, used internally by `#dup`, which makes a copy of `@parameters`.

Commit:
Changes to a dupped `ActionController::Parameters` mutate the original

When `ActionController::Parameters` is duplicated with `#dup`, it doesn't create a duplicate of the instance variables (e.g. `@parameters`) but rather maintains the reference (see <http://ruby-doc.org/core-2.3.1/Object.html>). Given that the parameters object is often manipulated as if it were a hash (e.g. with `#delete` and similar methods), this leads to unexpected behaviour, like the following:

```
params = ActionController::Parameters.new(foo: "bar")
duplicated_params = params.dup
duplicated_params.delete(:foo)

params == duplicated_params
```

This fixes the bug by defining a private `#initialize_copy` method, used internally by `#dup`, which makes a copy of `@parameters`.

bluehallu commented:
Isn't 

`new_params = params.except :foo`

The better approach?

timrogers commented:
@bluehallu Using `#except` is a partial solution, but to me the `#dup` behaviour seems pretty broken! Keen to see what @chancancode thinks though.

timrogers commented:
I must say, I'm confused by the intention of the existing `#dup` method - wouldn't `@permitted` get copied anyway?

matthewd commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -783,6 +783,11 @@ def hash_filter(params, filter)
+
+      def initialize_copy(source)
+        super
+        @parameters = source.instance_variable_get(:@parameters).dup
```
At this point, we can just use `@parameters = @parameters.dup`, I think?

matthewd commented:
> I must say, I'm confused by the intention of the existing #dup method - wouldn't @permitted get copied anyway?

I imagine that may not have been true while this class inherited from Hash.

If you can confirm it works the same without it (and that we have test coverage to back that up), this seems like a good opportunity to clean up, before we make things look even more complicated, with a custom `dup` *and* an `initialize_copy`.

Commit:
Trust `Object#dup` in `ActionController::Parameters`, using `#initialize_copy` to manually duplicate the underlying parameters hash

It looks like `ActionController::Parameters#dup` is leftover from when the class inherited from `Hash`. We can just trust `#dup`, which already copies the `@permitted` instance variable (confirmed by tests). We still define a `#initialize_copy` to make `@parameters` a copy that can be mutated without affecting the original instance.

timrogers commented:
@matthewd Agreed. Can you take another look? I've removed `#dup` altogether and replaced it with a simpler `#initialize_copy`. Some tests I've added confirm that `@permitted` is preserved.

matthewd commented:
Thanks! :heart:

rafaelfranca commented:
Backported in 702db0c

rafaelfranca commented:
Backported in 702db0c
